### DISCUSSION TITLE
[TestClient][HowTo] Design pytest for a feature that handshakes with an external service then receives a callback.

### DISCUSSION THREAD

**Discussion by Neah-Ko** [encode/starlette#2626] [Potential Issue] [answered]
Hello,

I'm building a [little project](https://github.com/bag-cnag/biodm/tree/dev) which leverages Starlette for ASGI, for my organization.

It includes a login feature which is essentially doing open-id token retrieval querying a [keycloak](https://www.keycloak.org/documentation ) instance that manages our keys.
The workflow is the following:
- User hits my `/login` server endpoint which returns a keycloak login url (including `redirect_uri` callback)
- User visits that url and logs-in on the form
- On success keycloak sends back an `authentication code` at that `redirect_uri`
- Server endpoint `/syn_ack` (callback target) receives that code and redeems it for a JW Token then returns it in a response.

I'm designing a test that covers the case:
```python
import json
import requests
from bs4 import BeautifulSoup

@pytest.fixture(scope="session", autouse=True)
def client():
    # Some shenanigans to import 'src/example' codebase from 'src/tests'.
    cwd = Path(__file__).parent
    example = cwd.parent.parent / 'example'
    sys.path.append(example)

    from example.app import main

    with TestClient(app=main(), backend_options={
        "use_uvloop": True
    }) as c:
        yield c

def json_bytes(d: Dict[Any, Any]) -> bytes:
    """Encodes python Dict as utf-8 bytes."""
    return json.dumps(d).encode('utf-8')

def test_login_user_on_keycloak_and_get_token(client):
    # Create User
    user = {"username": "u_test", "password": "1234"}
    _ = client.post('/users', content=json_bytes(user))

    # Get login page
    login_response = client.get('login')
    login_url = login_response.text

    with requests.Session() as session:
        # Get and Parse form with bs
        # Courtesy of: https://www.pythonrequests.com/python-requests-keycloak-login/
        form_response = session.get(login_url)
        soup = BeautifulSoup(form_response.content, 'html.parser')
        form = soup.find('form')
        action = form['action']
        other_fields = {i['name']: i.get('value', '') for i in form.findAll('input', {'type': 'hidden'})}

        response = session.post(action, data={
            'username': user['username'],
            'password': user['password'],
            **other_fields,
        }, allow_redirects=True)

    assert response.status_code == 201
```

However, the `post` request fails when keycloak is trying to reach that `redirect_uri` since it seems that `TestClient` doesn't actually exposes itself to receive requests from the outside. I unsuccessfully tried probing `http://testserver` but it's simply a mock, no ?

What would be the way to go in order to test such a feature ?

Best,

**Comment by adriangb** [comment]
You'll need to either run your app as a webserver e.g. via docker compose or, if the keycloak client accept it, pass in your own `requests.Session` or `httpx.{Async}Client`

**Reply by Neah-Ko** [reply]
Thanks,

Yeah it obviously works if I have my server running within a docker compose network before executing tests.

However, that way I cannot run a meaningful debugger session since the server sources are remote.

**Reply by adriangb** [reply, marked as answer]
Well you can't get something that only talks HTTP over a network to talk to TestClient in process. You either make your server talk over the network or you make the thing talk in-process, both of which I suggested above.
